File: templates/backend-node/src/utils/logger.ts
/**
 * Logger utility.
 * 
 * Centralized logging functionality.
 */

import { config } from '../config';

export enum LogLevel {
  DEBUG = 'DEBUG',
  INFO = 'INFO',
  WARN = 'WARN',
  ERROR = 'ERROR'
}

class Logger {
  private logLevel: LogLevel;
  
  constructor() {
    this.logLevel = config.logLevel as LogLevel || LogLevel.INFO;
  }
  
  private shouldLog(level: LogLevel): boolean {
    const levels = [LogLevel.DEBUG, LogLevel.INFO, LogLevel.WARN, LogLevel.ERROR];
    return levels.indexOf(level) >= levels.indexOf(this.logLevel);
  }
  
  debug(message: string, ...args: any[]): void {
    if (this.shouldLog(LogLevel.DEBUG)) {
      console.debug(`[DEBUG] ${message}`, ...args);
    }
  }
  
  info(message: string, ...args: any[]): void {
    if (this.shouldLog(LogLevel.INFO)) {
      console.info(`[INFO] ${message}`, ...args);
    }
  }
  
  warn(message: string, ...args: any[]): void {
    if (this.shouldLog(LogLevel.WARN)) {
      console.warn(`[WARN] ${message}`, ...args);
    }
  }
  
  error(message: string, ...args: any[]): void {
    if (this.shouldLog(LogLevel.ERROR)) {
      console.error(`[ERROR] ${message}`, ...args);
    }
  }
}

export const logger = new Logger();

File: templates/app-flutter/lib/core/utils/logger.dart
/// Logger utility.
///
/// Centralized logging. Use this instead of print() or debugPrint() directly.
/// In release builds, DEBUG level messages are suppressed automatically.
import 'package:flutter/foundation.dart';

enum LogLevel { debug, info, warn, error }

class Logger {
  final String _tag;

  const Logger([this._tag = 'App']);

  void debug(String message) => _log(LogLevel.debug, message);

  void info(String message) => _log(LogLevel.info, message);

  void warn(String message) => _log(LogLevel.warn, message);

  void error(String message, [Object? error, StackTrace? stackTrace]) {
    _log(LogLevel.error, message);
    if (error != null) _log(LogLevel.error, 'Error: $error');
    if (stackTrace != null) _log(LogLevel.error, stackTrace.toString());
  }

  void _log(LogLevel level, String message) {
    if (!kDebugMode && level == LogLevel.debug) return;
    debugPrint('[${level.name.toUpperCase()}][$_tag] $message');
  }
}

/// Global logger instance. Use directly or create named instances per class:
///   final _log = Logger('AuthRepository');
final logger = Logger();
